media_driver/linux/common/ddi/media_libva.cpp

```diff
@@ -6002,6 +6002,128 @@ DdiMedia_CopyInternal(
+    {
+        DDI_CHK_LESS((uint32_t)dst_obj->object.buffer_id, mediaCtx->pBufferHeap->uiAllocatedHeapElements, "Invalid copy dst buf_id", VA_STATUS_ERROR_INVALID_BUFFER);
+        PDDI_MEDIA_BUFFER dst_buffer = DdiMedia_GetBufferFromVABufferID(mediaCtx, dst_obj->object.buffer_id);
+        DDI_CHK_NULL(dst_buffer, "nullptr buffer", VA_STATUS_ERROR_INVALID_BUFFER);
+        DDI_CHK_NULL(dst_buffer->pGmmResourceInfo, "nullptr dst_buffer->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&dst, sizeof(dst));
+        DdiMedia_MediaBufferToMosResource(dst_buffer, &dst);
+    }
+
+    if (src_obj->obj_type == VACopyObjectSurface)
+    {
+        DDI_CHK_LESS((uint32_t)src_obj->object.surface_id, mediaCtx->pSurfaceHeap->uiAllocatedHeapElements, "copy_src", VA_STATUS_ERROR_INVALID_SURFACE);
+        PDDI_MEDIA_SURFACE src_surface = DdiMedia_GetSurfaceFromVASurfaceID(mediaCtx, src_obj->object.surface_id);
+        DDI_CHK_NULL(src_surface, "nullptr surface", VA_STATUS_ERROR_INVALID_SURFACE);
+        DDI_CHK_NULL(src_surface->pGmmResourceInfo, "nullptr src_surface->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&src, sizeof(src));
+        DdiMedia_MediaSurfaceToMosResource(src_surface, &src);
+    }
+    else if (src_obj->obj_type == VACopyObjectBuffer)
+    {
+        DDI_CHK_LESS((uint32_t)src_obj->object.buffer_id, mediaCtx->pBufferHeap->uiAllocatedHeapElements, "Invalid copy dst buf_id", VA_STATUS_ERROR_INVALID_BUFFER);
+        PDDI_MEDIA_BUFFER src_buffer = DdiMedia_GetBufferFromVABufferID(mediaCtx, src_obj->object.buffer_id);
+        DDI_CHK_NULL(src_buffer, "nullptr buffer", VA_STATUS_ERROR_INVALID_BUFFER);
+        DDI_CHK_NULL(src_buffer->pGmmResourceInfo, "nullptr src_buffer->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&src, sizeof(src));
+        DdiMedia_MediaBufferToMosResource(src_buffer, &src);
+    }
+
+    MOS_ZeroMemory(&mosCtx, sizeof(mosCtx));
+
+    mosCtx.bufmgr          = mediaCtx->pDrmBufMgr;
+    mosCtx.m_gpuContextMgr = mediaCtx->m_gpuContextMgr;
+    mosCtx.m_cmdBufMgr     = mediaCtx->m_cmdBufMgr;
+    mosCtx.fd              = mediaCtx->fd;
+    mosCtx.iDeviceId       = mediaCtx->iDeviceId;
+    mosCtx.SkuTable        = mediaCtx->SkuTable;
+    mosCtx.WaTable         = mediaCtx->WaTable;
+    mosCtx.gtSystemInfo    = *mediaCtx->pGtSystemInfo;
+    mosCtx.platform        = mediaCtx->platform;
+
+    mosCtx.ppMediaCopyState      = &mediaCtx->pMediaCopyState;
+    mosCtx.pfnMediaCopy          = mediaCtx->pfnMediaCopy;
+    mosCtx.gtSystemInfo          = *mediaCtx->pGtSystemInfo;
+    mosCtx.m_auxTableMgr         = mediaCtx->m_auxTableMgr;
+    mosCtx.pGmmClientContext     = mediaCtx->pGmmClientContext;
+
+    mosCtx.m_osDeviceContext     = mediaCtx->m_osDeviceContext;
+    mosCtx.m_apoMosEnabled       = mediaCtx->m_apoMosEnabled;
+
+    pCpDdiInterface = Create_DdiCpInterface(mosCtx);
```

Comment:
Create on each copy call? Can we avoid it?

Reply:
currently, as driver design. Mos maintain the same MediaCopy as DDI device. and copy is based on Gen specific. mediaCtx will be created during vadpy open, it is also per device. here is just pick up the device existed instance. not created new one.